{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import regex\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "import cfg"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Load restrictions sites and codon tools\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "outputs": [],
   "source": [
    "def load_restriction_sites():\n",
    "\n",
    "    re_seqs = []\n",
    "    for record in SeqIO.parse(cfg.RESTRICTSITES_FN, 'fasta'):\n",
    "        re_seqs.append(record.seq)\n",
    "        re_seqs.append(record.seq.reverse_complement())\n",
    "\n",
    "    return(set(re_seqs))\n",
    "\n",
    "def load_hs_codon_freq():\n",
    "    '''\n",
    "    Generate aa to codon dictionary.\n",
    "    '''\n",
    "\n",
    "    aa_codon_dt = pd.read_csv(cfg.DATA_DIR+'hs_codon_freq.csv')\n",
    "    aa_to_codon = dict(aa_codon_dt.loc[\n",
    "            aa_codon_dt.groupby('aa')['freq'].idxmax()][\n",
    "                    ['aa','codon']].to_dict('split')['data'])\n",
    "\n",
    "    return(aa_codon_dt, aa_to_codon)\n",
    "\n",
    "AA_CODON_DT, AA_TO_CODON = load_hs_codon_freq()\n",
    "\n",
    "def rand_codon(aa, orig=None):\n",
    "    '''\n",
    "    select random codon\n",
    "    based on hs genome freq\n",
    "    '''\n",
    "    codon_rows = AA_CODON_DT[AA_CODON_DT['aa'] == aa]\n",
    "    if orig:\n",
    "        codon_rows = codon_rows[codon_rows.codon != orig]\n",
    "        if not len(codon_rows):\n",
    "            #no other codons, can't use orig\n",
    "            return rand_codon(aa)\n",
    "    return(np.random.choice(\n",
    "            codon_rows.codon,\n",
    "            p = codon_rows.freq/sum(codon_rows.freq)))\n",
    "\n",
    "RE_SITES = load_restriction_sites()\n",
    "RE_REGEX = regex.compile('('+'|'.join([str(re).replace('N','.') for re in RE_SITES])+')')\n",
    "LIBRARY = pd.read_csv('out/results.csv')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## RE Check\n",
    "\n",
    "Load the dictionary of mutated sequences generated by the TCR_library program from a saved CSV file. I then iterate through\n",
    "the dataframe and extract the mutant DNA and AA sequence. These sequences are then fed to the remove_re_sites function.\n",
    "remove_re_sites searches for instances of the specified site (including overlapping sequences) using the finditer function\n",
    "provided in the regex module."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "outputs": [],
   "source": [
    "def remove_re_sites(\n",
    "        nt_seq, aa_seq,\n",
    "        pre=cfg.OLIGO_CHECK_PRE,\n",
    "        post=cfg.OLIGO_CHECK_POST,\n",
    "        re_regex= RE_REGEX,\n",
    "        v= False):\n",
    "\n",
    "    #assert(Seq.Seq(nt_seq).translate() == aa_seq)\n",
    "    re_matches = re_regex.finditer(pre+nt_seq+post, overlapped = 1)\n",
    "    replacements = 0\n",
    "    codon_indices_replaced = []\n",
    "    aa_replaced = []\n",
    "    old_codons = []\n",
    "    new_codons = []\n",
    "\n",
    "    for match in re_matches:\n",
    "\n",
    "        if v:\n",
    "            print('replacements:{}'.format(replacements))\n",
    "            print(pre.lower() + nt_seq + post.lower())\n",
    "            print('global match:'+str(match))\n",
    "\n",
    "        if replacements > 100:\n",
    "            raise Exception('too many codon replacement attempts') # something is up\n",
    "\n",
    "        # get span of nucleotide indices\n",
    "\n",
    "        nt_span = (\n",
    "                max([0, match.span()[0]- len(pre)]),\n",
    "                min([len(nt_seq), match.span()[1] - len(pre)]))\n",
    "\n",
    "        # expand to corresponding codon indices\n",
    "        aa_span = [i // 3 for i in nt_span]\n",
    "\n",
    "        # convert codon indices back to nucleotides\n",
    "        nt_rep_span = (aa_span[0]*3, aa_span[1]*3+3)\n",
    "\n",
    "        if v: print('spans:'+str(nt_span)+str(aa_span)+str(nt_rep_span))\n",
    "\n",
    "        # try replacing one codon at a time:\n",
    "        for codon_i in random.sample(\n",
    "                range(aa_span[0], aa_span[1]),\n",
    "                aa_span[1]-aa_span[0]):\n",
    "\n",
    "            old_codon = nt_seq[(codon_i*3):((codon_i*3)+3)]\n",
    "            if v: print(codon_i)\n",
    "            new_codon = rand_codon(aa_seq[codon_i], orig=old_codon)\n",
    "            if v: print(new_codon)\n",
    "\n",
    "            if v:\n",
    "                print('changing codon {}:{}({}) to {}({})'.format(\n",
    "                        codon_i,\n",
    "                        old_codon,\n",
    "                        aa_seq[codon_i],\n",
    "                        new_codon,\n",
    "                        Seq.Seq(new_codon).translate()))\n",
    "\n",
    "            new_nt_seq = (\n",
    "                    nt_seq[:(codon_i*3)] +\n",
    "                    new_codon +\n",
    "                    nt_seq[((codon_i*3)+3):])\n",
    "\n",
    "            # print('new seq: '+pre+new_nt_seq+post)\n",
    "            # test for same match as found originally\n",
    "            local_re_match = re_regex.search((pre+new_nt_seq+post)[\n",
    "                    match.span()[0]:match.span()[1]])\n",
    "            if v: print('local match: '+str(local_re_match))\n",
    "\n",
    "            # if no more match, we're done, accept new seq\n",
    "            if not local_re_match:\n",
    "                nt_seq = new_nt_seq\n",
    "                codon_indices_replaced.append(codon_i)\n",
    "                aa_replaced.append(aa_seq[codon_i])\n",
    "                old_codons.append(old_codon)\n",
    "                new_codons.append(new_codon)\n",
    "                break\n",
    "\n",
    "        replacements += 1\n",
    "        #redo search to see if we still have restriction sites\n",
    "\n",
    "        #Concerned that this I'm not updating the iterator with the new search?\n",
    "        re_matches = re_regex.finditer(pre+nt_seq+post, overlapped = 1)\n",
    "\n",
    "    assert(Seq.Seq(nt_seq).translate() == aa_seq), '{} ne {}'.format(\n",
    "            Seq.Seq(nt_seq).translate(),\n",
    "            aa_seq)\n",
    "\n",
    "    return(nt_seq, zip(\n",
    "            codon_indices_replaced,\n",
    "            aa_replaced,\n",
    "            old_codons,\n",
    "            new_codons))"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "replacements:0\n",
      "cactgcgaggatctgaaaaacgtgttccctcctgaagtggctgtctttgaaccatccgaggccgGGTCTCagatttcccatacccagaaagcaactctggtctgtctggccactggattctaccccgatcacgtggaactgtcttggtgggtgaacggcaaggaagtccattccggagtctctaccgaccctcagcccctcaaggagcagcctgctctcaacgattctcggtactgcctgtcatctcgactgagagtgtctgccaccttctggcagaaccctagaaaccactttcggtgtcaggtccagttttacggcctgagcgagaacgatgagtggacacaggatagagccaaacctgtgacacagattgtgagcgccgaggcttggggacgagccgattgtggcttcacatccgagtcttaccagcagggagtgctgtctgctacaatcctctacgaaagaggg\n",
      "global match:<regex.Match object; span=(64, 70), match='GGTCTC'>\n",
      "spans:(58, 64)[19, 21](57, 66)\n",
      "20\n",
      "ATC\n",
      "changing codon 20:TCT(I) to ATC(I)\n",
      "local match: None\n",
      "      position original_aa new_aa new_codon  \\\n",
      "0            0           L      *       NaN   \n",
      "1            0           L      M       ATG   \n",
      "2            0           L      I       ATC   \n",
      "3            0           L      I       ATT   \n",
      "4            0           L      V       GTG   \n",
      "...        ...         ...    ...       ...   \n",
      "1747       150           E      K       AAA   \n",
      "1748       150           E      A       GCC   \n",
      "1749       150           E      A       GCT   \n",
      "1750       150           E      H       CAC   \n",
      "1751       150           E      H       CAT   \n",
      "\n",
      "                                            new_dna_seq  \\\n",
      "0     gaggatctgaaaaacgtgttccctcctgaagtggctgtctttgaac...   \n",
      "1     ATGgaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "2     ATCgaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "3     ATTgaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "4     GTGgaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "...                                                 ...   \n",
      "1747  ctggaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "1748  ctggaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "1749  ctggaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "1750  ctggaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "1751  ctggaggatctgaaaaacgtgttccctcctgaagtggctgtctttg...   \n",
      "\n",
      "                                             new_aa_seq  \n",
      "0     EDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWW...  \n",
      "1     MEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "2     IEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "3     IEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "4     VEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "...                                                 ...  \n",
      "1747  LEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "1748  LEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "1749  LEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "1750  LEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "1751  LEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSW...  \n",
      "\n",
      "[1752 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "def screen_library_for_re_sites(library):\n",
    "    '''\n",
    "\n",
    "    :param library: CSV file of unscreened sequences\n",
    "    :return: CSV file of screened sequences\n",
    "    '''\n",
    "\n",
    "    screened_sequences = []\n",
    "    re_sites_removed = 0\n",
    "\n",
    "    for sequence in library.iterrows(): #Iterate through CSV and convert back into sequence dictionary\n",
    "\n",
    "        screened_sequence = {\n",
    "            'position': sequence[1][1],\n",
    "            'original_aa': sequence[1][2],\n",
    "            'new_aa': sequence[1][3],\n",
    "            'new_codon': sequence[1][4],\n",
    "            'new_dna_seq': sequence[1][5],\n",
    "            'new_aa_seq': sequence[1][6]\n",
    "        }\n",
    "\n",
    "        dna_sequence_pre = screened_sequence['new_dna_seq']\n",
    "        aa_sequence_pre = screened_sequence['new_aa_seq']\n",
    "\n",
    "        dna_sequence_post, re_info = remove_re_sites(dna_sequence_pre,aa_sequence_pre, v= True) #Check for RE sites\n",
    "\n",
    "        re_info = list(re_info)\n",
    "\n",
    "        if len(re_info) > 0: #Update the DNA sequence if a site if found\n",
    "            re_sites_removed += len(re_info[0])\n",
    "            screened_sequence ['new_dna_seq'] = dna_sequence_post\n",
    "\n",
    "        screened_sequences += [screened_sequence]\n",
    "\n",
    "\n",
    "    return pd.DataFrame(screened_sequences), re_sites_removed\n",
    "\n",
    "screened_library, re_sites = screen_library_for_re_sites(LIBRARY)\n",
    "\n",
    "screened_library.to_csv('out/screened_library.csv')\n",
    "print (screened_library)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}