{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# TCR Library Design"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [],
   "source": [
    "#Import reference sequence as DNA and AA seq objects\n",
    "\n",
    "REF_DNA = SeqIO.read(\"TRBC_Extracellular.fasta\", \"fasta\").seq\n",
    "REF_AA = Seq.translate(REF_DNA)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Setup:\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [],
   "source": [
    "BLOSSOM_SUB_TABLE = {} # Dictionary with list of substitutions for a given amino acid based on blossom matrix\n",
    "HOMOLOGY_SUB_TABLE = {} # Dictionary with list of mutation frequencies for a given position based on homologous variants\n",
    "MUTATED_SEQUENCES = [] # List of lists containing dictionary entries for each mutated sequence. Index of first set of lists = position in AA sequence"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Mutation Generation:\n",
    "\n",
    "Mutations are generated by iterating through the amino acid sequence of the protein. The first 4 substitutions for a\n",
    "given position are alanine and the top 3 results from the blossom table. I then plan to look for the most frequent\n",
    "mutation seen at that position in homologous sequences that is not included in the intial set of substitutions.\n",
    "\n",
    "This list of AA substititions will be fed to Dan's code to pick the top two codons for that AA, formatted into a\n",
    "dictionary that includes the position mutated, the original AA at that position, the new AA at that position, the\n",
    "new codon used, the new DNA sequence, and the new AA sequence.\n",
    "\n",
    "These dictionaries"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [],
   "source": [
    "def delete_position(aa_i):\n",
    "    dna_i = aa_i * 3\n",
    "    deleted_position = {\n",
    "        \"position\": aa_i,\n",
    "        \"original_aa\": REF_AA[aa_i],\n",
    "        \"new_aa\": \"*\",\n",
    "        \"new_codon\": \"\",\n",
    "        \"new_dna_seq\": REF_DNA[:dna_i] + REF_DNA[dna_i + 3:],\n",
    "        \"new_aa_seq\": REF_AA[:aa_i] + REF_AA[aa_i+1:]\n",
    "    }\n",
    "\n",
    "    return [deleted_position]\n",
    "\n",
    "def generate_subs(aa_i):\n",
    "    aa = REF_AA[aa_i] #AA being substituted\n",
    "    sub_aas = [] #List of new AAs\n",
    "    sub_sequences = [] #List of mutation dictionaries\n",
    "\n",
    "    blossom_subs = BLOSSOM_SUB_TABLE[aa] #Alanine + top 3 mutations from blossom table\n",
    "    homologous_subs = HOMOLOGY_SUB_TABLE[aa_i]\n",
    "\n",
    "    return sub_sequences\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Testing:\n",
    "Basic sanity checks for a given mutation:\n",
    "\n",
    "1.) Does the translated new DNA sequence match the new AA sequence?\n",
    "\n",
    "2.) Is the new DNA sequence different from the reference DNA sequence?\n",
    "\n",
    "3.) Does the new codon match the new AA?\n",
    "\n",
    "4.) Does the new AA match the"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [],
   "source": [
    "def check_translation(mutation_dict):\n",
    "    if Seq.translate(mutation_dict[\"new_dna_seq\"].seq) != mutation_dict[\"new_aa_seq\"]:\n",
    "        return False\n",
    "\n",
    "    return True\n",
    "\n",
    "def check_dna_change(mutation_dict):\n",
    "    if mutation_dict[\"new_dna_seq\"] == REF_DNA:\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "for aa_i in range(len(REF_AA)):\n",
    "    MUTATED_SEQUENCES += delete_position(aa_i)\n",
    "    \n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}