{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# TCR Library Design"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#import cfg\n",
    "\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "#Import reference sequence as DNA and AA seq objects\n",
    "\n",
    "REF_DNA = SeqIO.read(\"TRBC_Extracellular.fasta\", \"fasta\").seq\n",
    "REF_AA = Seq.translate(REF_DNA)\n",
    "\n",
    "BLOSSOM_SUB_TABLE = pd.read_csv('blosum80_subs_plus_a.csv') # Dataframe with list of substitutions for a given amino acid based on blossom matrix\n",
    "\n",
    "MUTATED_SEQUENCES = [] # List of lists containing dictionary entries for each mutated sequence. Index of first set of lists = position in AA sequence"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-13-0bf0e7820adb>:25: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.\n",
      "  top_2_codons_dt = aa_codon_dt.groupby(['aa'])['codon','freq'].apply(lambda x: x.nlargest(2, columns=['freq']))\n"
     ]
    }
   ],
   "source": [
    "def homologous_sub_freq(): #Generate dictionary with ranked list of substitutions at a given position\n",
    "    homology_dt = pd.read_csv('homologous_subs.csv') #DT with sub frequencies from alignment\n",
    "    sub_freq_dict = {} #Dictionary with pre-sorted list of subs at a given position\n",
    "\n",
    "    aa_row = list(homology_dt[\"AA\"]) #List of which row corresponds to which AA\n",
    "\n",
    "    for position in homology_dt.keys()[1:]: #Iterates through sequence but skips first column with AA table\n",
    "        position_subs = list(homology_dt[position]) #Frequencies of all subs (index matches aa_row)\n",
    "        sub_tuples = [] #List of formatted sub frequencies (sub aa, frequency)\n",
    "\n",
    "        for i in range(len(position_subs)): #Iterates through list of sub frequencies and reformats them into tuples\n",
    "\n",
    "            aa = aa_row[i] #sub aa\n",
    "            sub_freq = position_subs[i] #frequency in homologous sequences\n",
    "\n",
    "            sub_tuples += [(aa, sub_freq)]\n",
    "\n",
    "        subs_by_freq = sorted(sub_tuples, key=lambda x: x[1], reverse=True) #Sorted list of subs\n",
    "        sub_freq_dict[int(position)] = subs_by_freq\n",
    "\n",
    "    return sub_freq_dict\n",
    "\n",
    "def load_hs_codon_freq(): #Generate dictionary of top 2 codons for each AA\n",
    "    aa_codon_dt = pd.read_csv('hs_codon_freq.csv') #Dataframe of codon frequencies\n",
    "    top_2_codons_dt = aa_codon_dt.groupby(['aa'])['codon','freq'].apply(lambda x: x.nlargest(2, columns=['freq']))\n",
    "    aa_to_codon = {} #Dictionary with list of top two codons for a given AA\n",
    "\n",
    "    for codon in top_2_codons_dt.iterrows():\n",
    "        aa = codon[0][0]\n",
    "        codon_dna = codon[1][0]\n",
    "\n",
    "        if aa in aa_to_codon.keys():\n",
    "            aa_to_codon[aa] += [codon_dna]\n",
    "\n",
    "        else:\n",
    "            aa_to_codon[aa] = [codon_dna]\n",
    "\n",
    "    return(aa_to_codon)\n",
    "\n",
    "AA_TO_CODON = load_hs_codon_freq()\n",
    "HOMOLOGY_SUB_TABLE = homologous_sub_freq()\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Mutation Generation:\n",
    "\n",
    "Mutations are generated by iterating through the amino acid sequence of the protein. The first 4 substitutions for a\n",
    "given position are alanine and the top 3 results from the blossom table. I then plan to look for the most frequent\n",
    "mutation seen at that position in homologous sequences that is not included in the intial set of substitutions.\n",
    "\n",
    "This list of AA substititions will be fed to Dan's code to pick the top two codons for that AA, formatted into a\n",
    "dictionary that includes the position mutated, the original AA at that position, the new AA at that position, the\n",
    "new codon used, the new DNA sequence, and the new AA sequence.\n",
    "\n",
    "These dictionaries"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [],
   "source": [
    "def delete_position(aa_i):\n",
    "    dna_i = aa_i * 3\n",
    "\n",
    "    deleted_position = {\n",
    "        \"position\": aa_i,\n",
    "        \"original_aa\": REF_AA[aa_i],\n",
    "        \"new_aa\": \"*\",\n",
    "        \"new_codon\": \"\",\n",
    "        \"new_dna_seq\": REF_DNA[:dna_i] + REF_DNA[dna_i + 3:],\n",
    "        \"new_aa_seq\": REF_AA[:aa_i] + REF_AA[aa_i+1:]\n",
    "    }\n",
    "\n",
    "    return [deleted_position]\n",
    "\n",
    "def generate_subs(aa_i):\n",
    "    unique_subs_count = 0\n",
    "\n",
    "    dna_i = aa_i * 3\n",
    "    aa = REF_AA[aa_i] #AA being substituted\n",
    "\n",
    "    sub_aas = [] #List of new AAs\n",
    "    sub_sequences = [] #List of mutation dictionaries\n",
    "\n",
    "    blossom_subs = list(BLOSSOM_SUB_TABLE[aa]) #Alanine + top 4 mutations from blossom table\n",
    "    sub_aas += blossom_subs\n",
    "\n",
    "\n",
    "    if aa_i >= 2:\n",
    "        for sub, freq in HOMOLOGY_SUB_TABLE[aa_i]:\n",
    "            if sub not in sub_aas and freq > 0:\n",
    "                sub_aas += [sub]\n",
    "                unique_subs_count += 1\n",
    "\n",
    "\n",
    "    for sub_aa in sub_aas:\n",
    "        codons = AA_TO_CODON[sub_aa]\n",
    "\n",
    "        for codon in codons:\n",
    "            sub_sequence = {\n",
    "                \"position\": aa_i,\n",
    "                \"original_aa\": REF_AA[aa_i],\n",
    "                \"new_aa\": sub_aa,\n",
    "                \"new_codon\": codon,\n",
    "                \"new_dna_seq\": REF_DNA[:dna_i] + codon + REF_DNA[dna_i + 3:],\n",
    "                \"new_aa_seq\": REF_AA[:aa_i] + sub_aa + REF_AA[aa_i+1:]\n",
    "            }\n",
    "\n",
    "            sub_sequences += [sub_sequence]\n",
    "\n",
    "    return sub_sequences, unique_subs_count"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Testing:\n",
    "Basic sanity checks for a given mutation:\n",
    "\n",
    "1.) Does the translated new DNA sequence match the new AA sequence?\n",
    "\n",
    "2.) Is the new DNA sequence different from the reference DNA sequence?\n",
    "\n",
    "3.) Does the new codon match the new AA?\n",
    "\n",
    "4.) Does the new AA match the"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73\n",
      "1752\n"
     ]
    }
   ],
   "source": [
    "def check_translation(mutation_dict):\n",
    "    if Seq.translate(mutation_dict[\"new_dna_seq\"]) != mutation_dict[\"new_aa_seq\"]:\n",
    "        return False\n",
    "\n",
    "    return True\n",
    "\n",
    "def check_dna_change(mutation_dict):\n",
    "    if mutation_dict[\"new_dna_seq\"] == REF_DNA:\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "count = 0\n",
    "for aa_i in range(len(REF_AA)):\n",
    "    MUTATED_SEQUENCES += delete_position(aa_i)\n",
    "    MUTATED_SEQUENCES += generate_subs(aa_i)[0]\n",
    "\n",
    "    count += generate_subs(aa_i)[1]\n",
    "\n",
    "\n",
    "print (count)\n",
    "print((len(MUTATED_SEQUENCES)))\n",
    "\n",
    "#for sequence in MUTATED_SEQUENCES:\n",
    "#    print(check_dna_change(sequence))\n",
    "#    print(check_translation(sequence))\n",
    "\n",
    "pd.DataFrame(MUTATED_SEQUENCES).to_csv('/Users/chris/PycharmProjects/TCR_Library/results.csv')\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}