{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# TCR Library Design"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from Bio import Seq\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "outputs": [],
   "source": [
    "#Import reference sequence as DNA and AA seq objects\n",
    "\n",
    "REF_DNA = SeqIO.read(\"data/TRBC_Extracellular.fasta\", \"fasta\").seq\n",
    "REF_AA = Seq.translate(REF_DNA)\n",
    "\n",
    "REF_DNA = REF_DNA.upper()\n",
    "\n",
    "BLOSSOM_SUB_TABLE = pd.read_csv('data/blosum80_subs_plus_a.csv') # Dataframe with list of substitutions for a given amino acid based on blossom matrix"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-419-7b734b04adca>:25: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.\n",
      "  top_2_codons_dt = aa_codon_dt.groupby(['aa'])['codon','freq'].apply(lambda x: x.nlargest(2, columns=['freq']))\n"
     ]
    }
   ],
   "source": [
    "def homologous_sub_freq(): #Generate dictionary with ranked list of substitutions at a given position\n",
    "    homology_dt = pd.read_csv('data/homologous_subs.csv') #DT with sub frequencies from alignment\n",
    "    sub_freq_dict = {} #Dictionary with pre-sorted list of subs at a given position\n",
    "\n",
    "    aa_row = list(homology_dt[\"AA\"]) #List of which row corresponds to which AA\n",
    "\n",
    "    for position in homology_dt.keys()[1:]: #Iterates through sequence but skips first column with AA table\n",
    "        position_subs = list(homology_dt[position]) #Frequencies of all subs (index matches aa_row)\n",
    "        sub_tuples = [] #List of formatted sub frequencies (sub aa, frequency)\n",
    "\n",
    "        for i in range(len(position_subs)): #Iterates through list of sub frequencies and reformats them into tuples\n",
    "\n",
    "            aa = aa_row[i] #sub aa\n",
    "            sub_freq = position_subs[i] #frequency in homologous sequences\n",
    "\n",
    "            sub_tuples += [(aa, sub_freq)]\n",
    "\n",
    "        subs_by_freq = sorted(sub_tuples, key=lambda x: x[1], reverse=True) #Sorted list of subs\n",
    "        sub_freq_dict[int(position)] = subs_by_freq\n",
    "\n",
    "    return sub_freq_dict\n",
    "\n",
    "def load_top_codons(): #Generate dictionary of top 2 codons for each AA\n",
    "    aa_codon_dt = pd.read_csv('data/hs_codon_freq.csv') #Dataframe of codon frequencies\n",
    "    top_2_codons_dt = aa_codon_dt.groupby(['aa'])['codon','freq'].apply(lambda x: x.nlargest(2, columns=['freq']))\n",
    "    top_codons = {} #Dictionary with list of top two codons for a given AA\n",
    "\n",
    "    for codon in top_2_codons_dt.iterrows():\n",
    "        aa = codon[0][0]\n",
    "        codon_dna = codon[1][0]\n",
    "\n",
    "        if aa in top_codons.keys():\n",
    "            top_codons[aa] += [codon_dna]\n",
    "\n",
    "        else:\n",
    "            top_codons[aa] = [codon_dna]\n",
    "\n",
    "    return top_codons,aa_codon_dt\n",
    "\n",
    "TOP_2_CODONS, AA_CODON_DT = load_top_codons()\n",
    "HOMOLOGY_SUB_TABLE = homologous_sub_freq()\n",
    "\n",
    "def rand_codon(aa, orig=None):\n",
    "    '''\n",
    "    select random codon\n",
    "    based on hs genome freq\n",
    "    '''\n",
    "    codon_rows = AA_CODON_DT[AA_CODON_DT['aa'] == aa]\n",
    "    if orig:\n",
    "        codon_rows = codon_rows[codon_rows.codon != orig]\n",
    "        if not len(codon_rows):\n",
    "            #no other codons, can't use orig\n",
    "            return rand_codon(aa)\n",
    "    return(np.random.choice(\n",
    "            codon_rows.codon,\n",
    "            p = codon_rows.freq/sum(codon_rows.freq)))"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Mutation Generation:\n",
    "\n",
    "Mutations are generated by iterating through the amino acid sequence of the protein. The first 4 substitutions for a\n",
    "given position are alanine and the top 3 results from the blossom table. I then plan to look for the most frequent\n",
    "mutation seen at that position in homologous sequences that is not included in the intial set of substitutions.\n",
    "\n",
    "This list of AA substititions will be fed to Dan's code to pick the top two codons for that AA, formatted into a\n",
    "dictionary that includes the position mutated, the original AA at that position, the new AA at that position, the\n",
    "new codon used, the new DNA sequence, and the new AA sequence.\n",
    "\n",
    "These dictionaries"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "outputs": [],
   "source": [
    "def delete_position(ref_dna, ref_aa, aa_i):\n",
    "    '''\n",
    "    Delete the AA in ref_aa specified by aa_i\n",
    "    :param ref_dna:\n",
    "    :param ref_aa:\n",
    "    :param aa_i:\n",
    "    :return:\n",
    "    '''\n",
    "\n",
    "    dna_i = aa_i * 3\n",
    "\n",
    "    deleted_position = {\n",
    "        \"position\": aa_i,\n",
    "        \"original_aa\": ref_aa[aa_i],\n",
    "        \"original_codon\": ref_dna[dna_i:dna_i + 3],\n",
    "        \"new_aa\": \"DEL\",\n",
    "        \"new_codon\": \"\",\n",
    "        \"new_dna_seq\": ref_dna[:dna_i] + ref_dna[dna_i + 3:],\n",
    "        \"new_aa_seq\": ref_aa[:aa_i] + ref_aa[aa_i+1:],\n",
    "        \"dna_change_s\": dna_i,\n",
    "        \"dna_change_e\": dna_i + 2\n",
    "\n",
    "    }\n",
    "\n",
    "    return [deleted_position]\n",
    "\n",
    "def insert_stop(ref_dna, ref_aa, aa_i):\n",
    "    \"\"\"\n",
    "    Generates a stop codon at a given position in the input ref_dna sequence\n",
    "    :param ref_dna:\n",
    "    :param ref_aa:\n",
    "    :param aa_i:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dna_i = aa_i * 3\n",
    "    aa = ref_aa[aa_i] #AA being substituted\n",
    "    original_codon = ref_dna[dna_i:dna_i + 3]\n",
    "    new_codon = TOP_2_CODONS[\"*\"][0]\n",
    "\n",
    "    changed_dna_i = [dna_i + i for i in range(len(new_codon)) if original_codon[i] != new_codon[i]]\n",
    "\n",
    "    sub_sequence = {\n",
    "        \"position\": aa_i,\n",
    "        \"original_aa\": ref_aa[aa_i],\n",
    "        \"original_codon\": original_codon,\n",
    "        \"new_aa\": \"STOP\",\n",
    "        \"new_codon\": new_codon,\n",
    "        \"new_dna_seq\": ref_dna[:dna_i] + new_codon + ref_dna[dna_i + 3:],\n",
    "        \"new_aa_seq\": ref_aa[:aa_i] + '*' + ref_aa[aa_i+1:],\n",
    "        \"dna_change_s\": int(min(changed_dna_i)),\n",
    "        \"dna_change_e\": int(max(changed_dna_i))\n",
    "    }\n",
    "\n",
    "    return [sub_sequence]\n",
    "\n",
    "def generate_subs(ref_dna, ref_aa, aa_i):\n",
    "    '''\n",
    "    Generates substitute sequence dictionary entries for a given position to alanine, the top 4 blossom substitutions,\n",
    "    any potential substitutions found in homologous sequences, and a synonymous mutation.\n",
    "\n",
    "    :param ref_dna: DNA sequence being mutated\n",
    "    :param ref_aa:  AA sequence being mutated\n",
    "    :param aa_i: Position which is being mutated\n",
    "    :return: List of dictionary entries containing details of mutations\n",
    "    '''\n",
    "\n",
    "    unique_subs_count = 0\n",
    "\n",
    "    dna_i = aa_i * 3\n",
    "    aa = ref_aa[aa_i] #AA being substituted\n",
    "    original_codon = ref_dna[dna_i:dna_i + 3]\n",
    "\n",
    "\n",
    "    sub_aas = [] #List of new AAs\n",
    "    sub_sequences = [] #List of mutation dictionaries\n",
    "\n",
    "    blossom_subs = list(BLOSSOM_SUB_TABLE[aa]) #Alanine + top 4 mutations from blossom table\n",
    "    sub_aas += blossom_subs\n",
    "\n",
    "    #Add all homologous substitutions to list of subs\n",
    "    if aa_i >= 2: #Only perform after start codon\n",
    "        for sub, freq in HOMOLOGY_SUB_TABLE[aa_i]:\n",
    "\n",
    "            if sub not in sub_aas and freq > 0 and sub != aa:\n",
    "                sub_aas += [sub]\n",
    "                unique_subs_count += 1\n",
    "\n",
    "    #Iterate through list of sub AAs and generate 2 codons per sub\n",
    "    for sub_aa in sub_aas:\n",
    "        codons = TOP_2_CODONS[sub_aa]\n",
    "\n",
    "        for codon in codons:\n",
    "\n",
    "            #Identify location of changed DNA indexes for tiling\n",
    "            changed_dna_i = [dna_i + i for i in range(len(codon)) if original_codon[i] != codon[i]]\n",
    "\n",
    "            sub_sequence = {\n",
    "                \"position\": aa_i,\n",
    "                \"original_aa\": ref_aa[aa_i],\n",
    "                \"original_codon\": original_codon,\n",
    "                \"new_aa\": sub_aa,\n",
    "                \"new_codon\": codon,\n",
    "                \"new_dna_seq\": ref_dna[:dna_i] + codon + ref_dna[dna_i + 3:],\n",
    "                \"new_aa_seq\": ref_aa[:aa_i] + sub_aa + ref_aa[aa_i+1:],\n",
    "                \"dna_change_s\": int(min(changed_dna_i)),\n",
    "                \"dna_change_e\": int(max(changed_dna_i))\n",
    "            }\n",
    "\n",
    "            sub_sequences += [sub_sequence]\n",
    "\n",
    "    #Attempt to generate a synonymous mutation at the position\n",
    "    silent_sub = rand_codon(aa, orig=original_codon)\n",
    "\n",
    "    #Create a record if a synonymous mutation was able to be generated\n",
    "    if silent_sub != original_codon:\n",
    "\n",
    "        changed_dna_i = [dna_i + i for i in range(len(silent_sub)) if original_codon[i] != silent_sub[i]]\n",
    "\n",
    "        silent_sub_sequence = {\n",
    "                    \"position\": aa_i,\n",
    "                    \"original_aa\": ref_aa[aa_i],\n",
    "                    \"original_codon\": original_codon,\n",
    "                    \"new_aa\": aa,\n",
    "                    \"new_codon\": silent_sub,\n",
    "                    \"new_dna_seq\": ref_dna[:dna_i] + silent_sub + ref_dna[dna_i + 3:],\n",
    "                    \"new_aa_seq\": ref_aa,\n",
    "                    \"dna_change_s\": int(min(changed_dna_i)),\n",
    "                    \"dna_change_e\": int(max(changed_dna_i))\n",
    "                }\n",
    "\n",
    "        sub_sequences += [silent_sub_sequence]\n",
    "\n",
    "    return sub_sequences, unique_subs_count\n",
    "\n",
    "def generate_library(ref_dna, ref_aa):\n",
    "    '''\n",
    "\n",
    "    :param ref_dna:\n",
    "    :param ref_aa:\n",
    "    :return:\n",
    "    '''\n",
    "\n",
    "\n",
    "    count = 0\n",
    "    mutated_sequences = []\n",
    "\n",
    "    for i in range(len(REF_AA)):\n",
    "        mutated_sequences += delete_position(ref_dna, ref_aa, i)\n",
    "        mutated_sequences += generate_subs(ref_dna, ref_aa, i)[0]\n",
    "\n",
    "        if i % 10 == 0:\n",
    "            mutated_sequences += insert_stop(ref_dna,ref_aa,i)\n",
    "\n",
    "    count += generate_subs(ref_dna, ref_aa, i)[1]\n",
    "\n",
    "    return pd.DataFrame(mutated_sequences)\n",
    "\n",
    "MUTATED_SEQUENCES = generate_library(REF_DNA, REF_AA)\n",
    "MUTATED_SEQUENCES.to_csv('out/results.csv')\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Testing:\n",
    "Basic sanity checks for a given mutation:\n",
    "\n",
    "1.) Does the translated new DNA sequence match the new AA sequence?\n",
    "\n",
    "2.) Is the new DNA sequence different from the reference DNA sequence?\n",
    "\n",
    "3.) Does the new codon match the new AA?\n",
    "\n",
    "4.) Does the new AA match the"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trimming duplicates...\n",
      "Size of initial library: 1906\n",
      "Size of final library: 1903\n",
      "Number of sequences removed: 3\n",
      "Trimming complete!\n"
     ]
    }
   ],
   "source": [
    "def trim_duplicates(library):\n",
    "    library_size = len(library)\n",
    "    unique_sequences = set()\n",
    "    unique_count = 0\n",
    "\n",
    "    non_unique_sequence_info = []\n",
    "    unique_sequence_info = []\n",
    "\n",
    "    for sequence in library.iterrows():\n",
    "        unique_sequences.add(str(sequence[1][5]))\n",
    "        if len(unique_sequences) > unique_count:\n",
    "            unique_count += 1\n",
    "            unique_sequence_info += [sequence[1]]\n",
    "        else:\n",
    "            non_unique_sequence_info += [sequence[1]]\n",
    "\n",
    "    non_unique_sequence_info = pd.DataFrame(non_unique_sequence_info)\n",
    "    unique_sequence_info = pd.DataFrame(unique_sequence_info)\n",
    "    removed_count = library_size - unique_count\n",
    "\n",
    "    print(\"Trimming duplicates...\")\n",
    "    print(\"Size of initial library: \" + str(library_size))\n",
    "    print(\"Size of final library: \" + str(unique_count))\n",
    "    print(\"Number of sequences removed: \" + str(removed_count))\n",
    "    print(\"Trimming complete!\")\n",
    "\n",
    "    return unique_sequence_info\n",
    "\n",
    "TRIMMED_INITIAL_LIBRARY = trim_duplicates(MUTATED_SEQUENCES)\n",
    "\n",
    "TRIMMED_INITIAL_LIBRARY.to_csv('out/results.csv')\n",
    "\n",
    "\n",
    "def check_translation(mutation_dict):\n",
    "    if Seq.translate(mutation_dict[\"new_dna_seq\"]) != mutation_dict[\"new_aa_seq\"]:\n",
    "        return False\n",
    "\n",
    "    return True\n",
    "\n",
    "def check_dna_change(mutation_dict):\n",
    "    if mutation_dict[\"new_dna_seq\"] == REF_DNA:\n",
    "        return False\n",
    "    return True"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}